### DISCUSSION TITLE
How to leave blank space for homework solutions?

### DISCUSSION THREAD

**Discussion by d-morrison** [quarto-dev/quarto-cli#5823] [Q&A] [answered]
### Description

I'm trying to use quarto to generate homework assignments and corresponding solution keys in pdf format. I figured out how to hide the solutions using profiles, like this:

```
---
title: "Homework 2"
format: pdf
---

# Standard errors

## 

Write down the definition of the standard error of an estimator $\hat \theta$, using prose (do not use any math symbols other than $\hat\theta$ for this sub-section).

::: {.content-visible when-profile="solutions"}
\textbf{Solution:}

The standard error of an estimator is the standard deviation of the estimator.
:::

## 

Write down the definition of the standard error of an estimator $\hat \theta$, using mathematical notation.

::: {.content-visible when-profile="solutions"}
\textbf{Solution:}

$\text{se}(\hat{\theta}) = \sqrt{\textbf{E}  \left[  \left(\hat\theta - \textbf{E}\left[  \hat\theta  \right]  \right)^2  \right]}$
:::

```

Here's the output with the default profile:

[hw-example.pdf](https://github.com/quarto-dev/quarto-cli/files/11659235/hw-example.pdf)

Here's the output with `profile="solutions"`:

[hw-example-SOLUTIONS.pdf](https://github.com/quarto-dev/quarto-cli/files/11659238/hw-example-SOLUTIONS.pdf)

I'm wondering if there's a way to leave blank space for students to write in their solutions, when rendering the default profile (without solutions).

Two solutions that I've tried are:

1. to have a div of blank lines that are `hidden-when profile="solutions"`. However, this requires me to manually adjust the number of blank lines to match the solutions. Ideally the amount of space in the default profile would be determined by the length of the hidden solutions, so that page numbers line up between the two profiles.

2. always print the solution text, but change the font color to white, unless profile="solutions". However, this approach means that the solutions are recoverable in dark mode, etc.

Thanks for any suggestions!

**Comment by mcanouil** [comment]
If you want dynamic height, you'll have to compute that using code cells and output LaTeX code, it seems harder than actually needed.

See below an example.
You can always add page breaks to keep "similar" page layout.
I am using `\vspace` but you can use whatever LaTeX command.

(There is some ongoing work to allow better conditional, such as if "profile" then "something" else "something else")

<table>
<tr><th>Quarto document</th><th>PDF</th></tr>
<tr>
<td>

````qmd
---
title: "Homework 2"
format: pdf
---

# Standard errors

## Question 1

Write down the definition of the standard error of an estimator $\hat \theta$,
using prose (do not use any math symbols other than $\hat\theta$ for this sub-section).

__Solution:__

::: {.content-visible when-profile="solutions"}
The standard error of an estimator is the standard deviation of the estimator.
:::

::: {.content-visible unless-profile="solutions"}
\vspace{1in}
:::

## Question 2

Write down the definition of the standard error
of an estimator $\hat \theta$, using mathematical notation.

__Solution:__

::: {.content-visible when-profile="solutions"}
$$
\text{se}(\hat{\theta}) = \sqrt{
  \textbf{E} \left[
    \left(\hat\theta - \textbf{E}\left[  \hat\theta  \right]  \right)^2
  \right]
}
$$
:::

::: {.content-visible unless-profile="solutions"}
\vspace{1in}
:::

````

</td>
<td>
<img width="686" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/39fdc0f7-920c-4572-82b6-a9931d781ff6">
</td>
</tr>
</table>

**Reply by d-morrison** [reply]
Thanks! That seems workable.

**Comment by cscheid** [comment]
> always print the solution text, but change the font color to white, unless profile="solutions". However, this approach means that the solutions are recoverable in dark mode, etc.

You could create a filter that replaces the strings of the solution text with random text of the same length, and then set the font color to your background color. It's not going to be perfectly equal, but it's probably good enough.

**Reply by d-morrison** [reply]
Thanks! This seems like a potential long-term solution, although I'll have to learn about filters first; I'm guessing you meant this: https://quarto.org/docs/extensions/filters.html ?

**Reply by cscheid** [reply]
Yes, that's what I meant. Lua filters give you the ability to manipulate the content of a document by writing code.

**Comment by baptiste** [comment, marked as answer]
Kind of a [weird LaTeX hack](https://tex.stackexchange.com/a/434156/269911), but tcolorbox has an "invisible" setting to not display the content, so you could have


````
---
title: "Homework 2"
format: 
   pdf:
     pdf-engine: xelatex
     latex-tinytex: false
     keep-tex: true
     keep-md: true
     include-in-header: 
       text: |
          \usepackage[most]{tcolorbox}
          
          \NewTColorBox[]{exercise}{+O{}}{%
                  enhanced, breakable,colframe=white,colback=white,
                  boxrule=0pt,frame hidden,upperbox=invisible,
                  #1}
---

## Standard errors

Write down the definition of the standard error of an estimator $\hat \theta$, using mathematical notation.
    
::: {.content-visible when-profile="exercise"}
`\begin{exercise}`{=latex}
:::

\textbf{Solution:}

Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. 

$\text{se}(\hat{\theta}) = \sqrt{\textbf{E}  \left[  \left(\hat\theta - \textbf{E}\left[  \hat\theta  \right]  \right)^2  \right]}$

Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.

::: {.content-visible when-profile="exercise"}
`\end{exercise}`{=latex}
:::

Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
````

> `quarto render test.qmd`

![Screenshot 2023-06-08 at 22 14 19](https://github.com/quarto-dev/quarto-cli/assets/18970/2755b0cd-c14b-4794-a49c-d3ea9137cf5a)


> `quarto render test.qmd --profile exercise`

![Screenshot 2023-06-08 at 22 13 33](https://github.com/quarto-dev/quarto-cli/assets/18970/caa32aa4-3349-4bd8-9290-edecf6806842)

**Reply by d-morrison** [reply]
Thanks! This seems like a complete solution for this use-case.

**Comment by d-morrison** [comment]
Is there a way to customize the "solution" environment (https://quarto.org/docs/authoring/cross-references.html#theorems-and-proofs) to do something similar? That might simplify the process of tagging the solution sections for hiding.

**Comment by cderv** [comment]
Maybe the **unilur** extension can help with this: https://github.com/ginolhac/unilur

**Comment by jgostick** [comment]
FWIW, I was looking for a solution that worked with HTML format, and the above did not cut it, so I used this:

```

<div style="padding-bottom:10cm; border:#E5E4E2; border-width:2px; border-style:solid;">

::: {.content-hidden when-format="html"}

- text a

- text b

- text c

:::

</div>

```
